File: Más Ejemplos/OMP-solution/pitfalls/race1.c

#include <stdio.h>

#ifdef _OPENMP
#include <omp.h>
#endif

double work1(int a){
  return a+1;
}

double work2(double a, int b){
  return a+b;
}

int main(void){
  double x,y,tmp;
  int i;

#pragma omp parallel shared(x) private(tmp) 
  {
#   pragma omp for reduction(+:x) 
    for(i=1; i< 100; i++){
      tmp=work1(i);
      x=x+tmp;
    }

#pragma omp single
    {
      y = x;
    }
  }/* end parallel*/
  printf("%g\n",y);
}

File: MasEjemplos/OMP-solution/pitfalls/race1.f90

real function WORK1(A)
integer A

WORK1=real(A+1)
end function WORK1


real function WORK2(A,B)
real A
integer B

WORK2=A+real(B)
end function WORK2

program race1
implicit none


real WORK1, WORK2
real X, TMP
real Y
integer I

       X=0
!$OMP PARALLEL SHARED (X), PRIVATE(TMP) 

!$OMP DO REDUCTION(+:X)
       DO 100 I=1,100
              TMP = WORK1(I)
              X = X + TMP
100  CONTINUE
!$OMP END DO 

!$OMP SINGLE
       Y = X
!$OMP END SINGLE
!$OMP END PARALLEL   

print *,Y



end program race1
